### PR TITLE
ARM system linkage improvements

### PR DESCRIPTION
**Pull request by Leonardo2718**
This series of commits improve/fix the ARM code generator in the OMR compiler. The main goal of this change is to complete the system linkage implementation. System linkage is the part of the code generator responsible for generating the instruction sequences at function call boundaries to satisfy the platform ABI.

These commits can be roughly grouped into 3 sections:

1. The first 6 commits do some necessary clean up to enable the other commits.
2. The next 4 commits constitute the main changes. These generalize parts of the ARM binary encoding and add missing functionality to the system linkage implementation.
3. The last 2 commits add a test case for the ARM system linkage and some trace in the code generator that was useful when debugging the previous changes. The test case is intended to exercises some edge cases in the system linkage implementation.

I am marking as WIP so that the changes can be reviewed while I run some more tests.

Issue #793

### CODE REVIEW

File: compiler/arm/codegen/ARMSystemLinkage.cpp [hunk 1]
```diff
@@ -167,156 +172,270 @@ TR::ARMLinkageProperties TR::ARMSystemLinkage::properties =
...
+   reg = machine->getARMRealRegister(TR::RealRegister::gr13);
+   reg->setState(TR::RealRegister::Locked);
+   reg->setAssignedRegister(reg);
+
+   // make r12 (IP) unavailable for RA
+   reg = machine->getARMRealRegister(TR::RealRegister::gr12);
+   reg->setState(TR::RealRegister::Locked);
+   reg->setAssignedRegister(reg);
+
+   // make r9 unavailable for RA (just in case, because it's meaning is platform defined)
```
Comment (0xdaryl) [inline compiler/arm/codegen/ARMSystemLinkage.cpp]: In the prologue/epilogue code below, r9 is treated as a preserved register.  If that's the case, I'm wondering why it isn't eligible for register assignment?

File: compiler/arm/codegen/ARMSystemLinkage.cpp [hunk 2]
```diff
@@ -167,156 +172,270 @@ TR::ARMLinkageProperties TR::ARMSystemLinkage::properties =
...
+               nextFltArgReg += 1;
+               mapSingleParameter(parameter, stackIndex);
+               }
+            else
+               {
+               nextFltArgReg = getProperties().getNumFloatArgRegs() + 1;
+               }
+            break;
+         case TR::Aggregate:
+            TR_ASSERT(false, "Function parameters of aggretate types are not currently supported on ARM.");
```
Comment (0xdaryl) [inline compiler/arm/codegen/ARMSystemLinkage.cpp]: aggretate -> aggregate

File: compiler/arm/codegen/ARMSystemLinkage.cpp (hunk 1 above)
Comment (Leonardo2718) [inline compiler/arm/codegen/ARMSystemLinkage.cpp]: According to the [AAPCS](http://infocenter.arm.com/help/topic/com.arm.doc.ihi0042f/IHI0042F_aapcs.pdf) (section 5.1.1), the meaning of r9 is platform dependent. Since it is unclear how the register is used on platforms targeted by downstream projects, I chose to do what seemed safest and made it both preserved and unavailable for register assignment.

File: compiler/arm/codegen/ARMSystemLinkage.cpp [hunk 3]
```diff
@@ -326,12 +445,122 @@ TR::ARMLinkageProperties& TR::ARMSystemLinkage::getProperties()
 
 void TR::ARMSystemLinkage::createPrologue(TR::Instruction *cursor)
    {
-   TR_ASSERT(0, "unimplemented");
+   TR::CodeGenerator   *codeGen    = cg();
+   const TR::ARMLinkageProperties& properties = getProperties();
+   TR::Machine         *machine    = codeGen->machine();
+   TR::ResolvedMethodSymbol* bodySymbol = comp()->getJittedMethodSymbol();
+   TR::Node                 *firstNode  = comp()->getStartTree()->getNode();
+   TR::RealRegister    *stackPtr   = machine->getARMRealRegister(properties.getStackPointerRegister());
```
Comment (Leonardo2718) [inline compiler/arm/codegen/ARMSystemLinkage.cpp]: This code actually came from a downstream project. I will fix.

**Comment by Leonardo2718** [comment]
@0xdaryl have I addressed all your concerns?

File: compiler/arm/codegen/ARMSystemLinkage.cpp (hunk 1 above)
Comment (lmaisons) [inline compiler/arm/codegen/ARMSystemLinkage.cpp]: >the meaning of r9 is platform dependent

Does this imply that there are asynchronous contexts where, despite using r9, we could be violating the expectation of the platform?

I'm having a hard time picturing the scenario that intersects both "requires preservation" and "don't touch this"

Comment (Leonardo2718) [inline compiler/arm/codegen/ARMSystemLinkage.cpp]: Not sure.

I'm open to changing this if someone can tell me which is the preferred alternative (between preserved vs unavailable for register assignment).

Comment (0xdaryl) [inline compiler/arm/codegen/ARMSystemLinkage.cpp]: Not being an ARM environment expert, is there any way to determine how r9 is intended to be used (e.g., build environment variable, runtime query, etc.) based on the "platform" (in ARM parlance) we're running in?  It would be a shame to default to the most conservative position (preserved and not available for RA) all the time when I suspect it actually is available in most cases.

What does the C runtime expect?

Comment (Leonardo2718) [inline compiler/arm/codegen/ARMSystemLinkage.cpp]: After researching the supported platform, there does not appear to be any straightforward documentation of how r9 is used. I don't really see any solution being viable without more detailed information about the target platforms, so I still think the existing implementation is safest for the moment.
